apps/test/JSInterpreterTest.js

```diff
@@ -0,0 +1,64 @@
+var testUtils = require('./util/testUtils');
+var assert = testUtils.assert;
+
+describe("JSInterpreter", function () {
+  var JSInterpreter = require('@cdo/apps/JSInterpreter');
+  var jsInterpreter;
+
+  // Setup up window.Interpreter.
+  window.acorn = require('../lib/jsinterpreter/acorn');
+  require('../lib/jsinterpreter/interpreter');
+
+  function assertCurrentState(expected) {
+    var state = jsInterpreter.interpreter.stateStack[0];
+    assert.containSubset(state, expected);
+  }
+
+  function stepAndVerify(expected) {
+    jsInterpreter.nextStep = JSInterpreter.StepType.IN;
+    jsInterpreter.executeInterpreter();
+    assertCurrentState(expected);
+  }
+
+  function verifyStepSequence(expectedStates) {
+    expectedStates.forEach(function (expected) {
+      stepAndVerify(expected);
+    });
+  }
+
+  it("steps a `for` loop", function () {
+
+    // Setup a jsInterpreter instance with `hideSource: true` so an editor isn't
+    // needed. Mock `getUserCodeLine` so the code is considered inUserCode.
+    jsInterpreter = new JSInterpreter({
+      studioApp: {hideSource: true}
+    });
+    jsInterpreter.getUserCodeLine = function () { return 0; };
+
+    // Initialize a test program
+    jsInterpreter.parse({
+      code: 'for (var i = 0; i < 2; i++) { 1; }'
+    });
+
+    assert(jsInterpreter.initialized());
+
+    // Step in
+    jsInterpreter.paused = true;
+    jsInterpreter.nextStep = JSInterpreter.StepType.IN;
+    jsInterpreter.executeInterpreter(true);
+
+    assertCurrentState({node: {type: 'ForStatement'}, mode: undefined});
+
+    // Continue stepping
+    verifyStepSequence([
+      {node: {type: 'ForStatement'}, mode: 1}, // (test) i < 2;
+      {node: {type: 'ExpressionStatement'}},   // (body) 1;
+      {node: {type: 'ForStatement'}, mode: 3}, // (update) i++
+      {node: {type: 'ForStatement'}, mode: 1}, // (test) i < 2;
+      {node: {type: 'ExpressionStatement'}},   // (body) 1;
+      {node: {type: 'ForStatement'}, mode: 3}, // (update) i++
+      {node: {type: 'ForStatement'}, mode: 1}  // (test) i < 2;
```

Comment:
It might be nice, even just for this test, to declare those as constants
for readability.

On Fri, Feb 12, 2016 at 5:44 PM, Josh Lory <notifications@github.com> wrote:

> In apps/test/JSInterpreterTest.js
> <https://github.com/code-dot-org/code-dot-org/pull/6784#discussion_r52820282>
> :
>
> > +    // Step in
> > +    jsInterpreter.paused = true;
> > +    jsInterpreter.nextStep = JSInterpreter.StepType.IN;
> > +    jsInterpreter.executeInterpreter(true);
> > +
> > +    assertCurrentState({node: {type: 'ForStatement'}, mode: undefined});
> > +
> > +    // Continue stepping
> > +    verifyStepSequence([
> > +      {node: {type: 'ForStatement'}, mode: 1}, // (test) i < 2;
> > +      {node: {type: 'ExpressionStatement'}},   // (body) 1;
> > +      {node: {type: 'ForStatement'}, mode: 3}, // (update) i++
> > +      {node: {type: 'ForStatement'}, mode: 1}, // (test) i < 2;
> > +      {node: {type: 'ExpressionStatement'}},   // (body) 1;
> > +      {node: {type: 'ForStatement'}, mode: 3}, // (update) i++
> > +      {node: {type: 'ForStatement'}, mode: 1}  // (test) i < 2;
>
> It describes which part of the ForStatement is being executed:
> mode sub-node
> 0 init
> 1 test
> 2 body
> 3 update
>
>
> https://github.com/NeilFraser/JS-Interpreter/blob/9c09d45c00d571be59b62319d2100658b728ce0f/interpreter.js#L2044-L2073
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/6784/files#r52820282>.
>



-- 
Brad Buchanan
Software Engineer, Code.org
brad@code.org